crates/runtime/src/memory.rs

```diff
@@ -126,13 +155,33 @@ impl MmapMemory {
             mmap.make_accessible(pre_guard_bytes, minimum)?;
         }
 
+        // If a memfd image was specified, try to create the MemFdSlot on top of our mmap.
+        let memfd = match memfd_image {
+            Some(image) => {
+                let base = unsafe { mmap.as_mut_ptr().offset(pre_guard_bytes as isize) };
+                let len = request_bytes - pre_guard_bytes;
+                let mut memfd_slot = MemFdSlot::create(base as *mut _, len);
+                memfd_slot.instantiate(minimum, Some(image))?;
```

Comment:
I've updated this so that the extra mmap no longer occurs. The basic idea is actually a slightly more general optimization: if there is no `image` (this can happen any time a memory with no data segments uses a `MemFdSlot`), then the whole slot is just the anon-zeroes mmap, and we don't need to wipe it clear with a new one. This is true when the `MemFdSlot` is first created as well: it has an `image` of `None` and we require the caller to have created a backing mmap already (as both the dynamic linear memory and the pooling allocator do).